### DISCUSSION TITLE
JPG to sRGB values WITHOUT substituting color names?

### DISCUSSION THREAD

**Discussion by kitfoxer** [ImageMagick/ImageMagick#2650] [Help] [answered]
Hello all, new to this forum and pretty new to IM...

I've been using ImageMagick to convert JPGs to their individual pixel values so that I can do some profile plotting.  Many of the pictures I analyze have neutral grays from photographic standards etc, so when I convert them the resulting file contains color names like "grey19" for a given pixel instead of the actual RGB values I am seeking.  So I then have to look up the names to get those values which is a nuisance.  Is there any way around this issue or is that just the way the conversion works?  Thanks in advance for any help.

**Comment by snibgo** [comment]
Please show your command and text output. Otherwise we can only guess.

Also say what version of IM, on what platform.

**Reply by kitfoxer** [reply]
I am on Win10, using IM version 7.0.8-7 Q16 x64 2018-07-17.
IM command: convert -depth 8 file.jpg file.txt
Sample output:
# ImageMagick pixel enumeration: 1760,53,65535,srgb
.
.
.
1656,6: (12079,12079,12079)  #2F2F2F  srgb(47,47,47)
1657,6: (12079,12079,12079)  #2F2F2F  srgb(47,47,47)
1658,6: (12079,12079,12079)  #2F2F2F  srgb(47,47,47)
1659,6: (12336,12336,12336)  #303030  grey19
1660,6: (12336,12336,12336)  #303030  grey19
1661,6: (12593,12593,12593)  #313131  srgb(49,49,49)
etc.
Thanks!

**Comment by snibgo** [comment]
I suggest you ignore everything after the hash "#".

The important numbers are those inside the parentheses, eg "(12079,12079,12079)". (These may be floating-point numbers.) If you prefer percentages, set a compliance, like this:
```
magick xc:sRGB(1.1010%,20.2020%,30.333%) -define txt:compliance=svg txt:

# ImageMagick pixel enumeration: 1,1,65535,srgb
0,0: (1.101%,20.202%,30.333%)  #02D233B74DA7  srgb(1.101%,20.202%,30.333%)
```

**Comment by fmw42** [comment]
The define does not seems to work for a named color.

```
magick xc:red -depth 8 -define txt:compliance=svg txt:
# ImageMagick pixel enumeration: 1,1,255,srgb
0,0: (255,0,0)  #FF0000  red

```


Add -depth 8 and then use the numbers in parenthesis. Those values will be the r,g,b values without rgb preface.

```
magick xc:red -depth 8 txt:
# ImageMagick pixel enumeration: 1,1,65535,srgb
0,0: (255,0,0)  #FF0000  red
```

**Comment by snibgo** [comment]
fmw42 wrote:
> The define does not seems to work for a named color.

It works fine for v7.0.8-64:
```
f:\web\im>%IMG7%magick xc:Red -define txt:compliance=svg  txt:

# ImageMagick pixel enumeration: 1,1,65535,srgb
0,0: (100%,0%,0%)  #FFFF00000000  red
```

... but fails for v7.0.10-27:
```
f:\web\im>%IMG701010%magick xc:Red -define txt:compliance=svg  txt:

# ImageMagick pixel enumeration: 1,1,65535,srgb
0,0: (255,0,0)  #FFFF00000000  red
```
Sigh.

**Reply by fmw42** [reply]
I was referring to the "red" at the end of the line not showing as rgb(255,0,0).

**Comment by snibgo** [comment]
I don't know of any way to change "red" to rgb(255,0,0) or similar.

**Reply by fmw42** [reply]
I believe that that field represent the color as conforming to the CSS guide. So you get color names. The -define txt:XXX was introduced to get different representations for the values in parentheses.

**Reply by kitfoxer** [reply]
Thanks everyone.  If I'm understanding you all correctly it looks like I will have to continue to cross-reference color names to RGB values.  I was kind of hoping there was the IM equivalent convert flag to doing an nslookup for an IP address based on a domain name to mix metaphors a bit!

**Comment by fmw42** [comment]
Why can you not use the values in parenthesis. They are the RGB values if you put in -depth 8.  There is a color converter at https://imagemagick.org/script/color.php and also a table of colors with their names and rgb values.  What is that you want to do?

```
magick xc:red -depth 8 -format "%[pixel:u.p]" info:
srgb(255,0,0)

```

**Reply by kitfoxer** [reply]
I want to take any JPG image (e.g. a 53-row crop about 1800 pixels wide) and convert each pixel in the row(s) to their RGB values.  Having IM insert color names is actually an impediment.  I want the RGB data, not the names.  Thanks for your help with syntax, I'm not a very experienced IM user at this point.

**Comment by fmw42** [comment, marked as answer]
`magick rose:[10x10+0+0] sparse-color: | tr " " "\n"`

```
0,0,srgb(48,47,45)
1,0,srgb(50,48,46)
2,0,srgb(54,50,47)
3,0,srgb(56,51,46)
4,0,srgb(58,51,45)
5,0,srgb(57,50,45)
6,0,srgb(56,48,45)
7,0,srgb(57,49,46)
8,0,srgb(56,48,45)
9,0,srgb(56,48,45)
0,1,srgb(47,46,44)
1,1,srgb(48,47,45)
2,1,srgb(53,48,45)
3,1,srgb(55,49,45)
4,1,srgb(56,49,44)
5,1,srgb(56,48,44)
6,1,srgb(54,47,44)
7,1,srgb(56,49,45)
8,1,srgb(56,49,45)
9,1,srgb(56,49,46)
0,2,srgb(44,44,44)
1,2,srgb(44,45,43)
2,2,srgb(48,47,44)
3,2,srgb(49,46,43)
4,2,srgb(50,44,40)
5,2,srgb(50,45,42)
6,2,srgb(48,44,43)
7,2,srgb(50,44,43)
8,2,srgb(50,43,43)
9,2,srgb(50,44,43)
0,3,srgb(44,45,47)
1,3,srgb(43,45,47)
2,3,srgb(46,45,45)
3,3,srgb(47,45,41)
4,3,srgb(47,42,39)
5,3,srgb(45,43,40)
6,3,srgb(44,42,42)
7,3,srgb(44,42,42)
8,3,srgb(46,43,43)
9,3,srgb(46,43,43)
0,4,srgb(44,47,50)
1,4,srgb(43,47,50)
2,4,srgb(46,46,48)
3,4,srgb(45,46,45)
4,4,srgb(45,43,41)
5,4,srgb(44,42,43)
6,4,srgb(43,41,43)
7,4,srgb(40,39,40)
8,4,srgb(42,42,43)
9,4,srgb(44,42,43)
0,5,srgb(46,48,46)
1,5,srgb(47,47,45)
2,5,srgb(47,49,46)
3,5,srgb(50,49,46)
4,5,srgb(51,48,45)
5,5,srgb(51,47,44)
6,5,srgb(50,45,42)
7,5,srgb(45,42,43)
8,5,srgb(44,43,44)
9,5,srgb(42,41,45)
0,6,srgb(50,51,45)
1,6,srgb(51,52,45)
2,6,srgb(53,51,46)
3,6,srgb(57,54,49)
4,6,srgb(57,54,49)
5,6,srgb(58,50,46)
6,6,srgb(56,48,45)
7,6,srgb(51,46,44)
8,6,srgb(51,48,48)
9,6,srgb(45,46,48)
0,7,srgb(55,56,50)
1,7,srgb(55,56,50)
2,7,srgb(57,57,52)
3,7,srgb(60,58,53)
4,7,srgb(64,61,56)
5,7,srgb(66,58,56)
6,7,srgb(65,57,53)
7,7,srgb(61,55,51)
8,7,srgb(60,55,53)
9,7,srgb(58,53,52)
0,8,srgb(56,57,51)
1,8,srgb(57,58,52)
2,8,srgb(62,62,56)
3,8,srgb(65,63,58)
4,8,srgb(68,63,58)
5,8,srgb(70,65,60)
6,8,srgb(69,65,58)
7,8,srgb(69,64,58)
8,8,srgb(71,63,56)
9,8,srgb(69,61,52)
0,9,srgb(54,55,49)
1,9,srgb(54,56,49)
2,9,srgb(61,61,56)
3,9,srgb(66,64,59)
4,9,srgb(70,65,63)
5,9,srgb(72,67,62)
6,9,srgb(72,67,62)
7,9,srgb(74,68,60)
8,9,srgb(77,66,55)
9,9,srgb(74,61,48)
```

"rose:" is an Internal ImageMagick image. You can replace rose:  with any "image.suffix"  (without the colon)

**Reply by kitfoxer** [reply]
This looks promising, but my system is reporting "'tr' is not recognized as an internal or external command, operable program or batch file."  So everything past that pipe is being refused.  Even when pasted as in your example above. I'm using a DOS command window for running IM commands if that makes any difference.  Win 10/64 bit.

**Reply by kitfoxer** [reply]
It looks like a simple substitution of a line feed for each space, but I'm not familiar with the "tr" part of it.

**Reply by fmw42** [reply]
tr is a unix command. But you can download a Windows version of it. Sorry, I am not a Windows user. But search Google for tr for Windows or unix tools for windows.

The tr is replacing a space with a line feed. You can leave off the | tr ... and still get the listing, but it will be one continuous string with spaces between the colors.

**Comment by snibgo** [comment]
>  I want the RGB data, not the names.

But you have the RGB data, in the example you showed:
```
1656,6: (12079,12079,12079) #2F2F2F srgb(47,47,47)
```
The RGB data is 12079,12079,12079, on a scale of 0 to 65535.

**Reply by kitfoxer** [reply]
We use 8-bit color for most of our purposes.  It's simpler and suffices for what we want in my case.  The command that fmw42 supplied works if I skip the pipe and open the file in Notepad++ to insert the line breaks.  Beats translating color names back to RGB values.  Time to drag out some ancient DOS stuff and see if I can make the substitution work on the command line.  Thanks everyone!

**Reply by fmw42** [reply]
If you add -depth 8 to the command, then the numbers in the parentheses will be 8-bit values.

**Reply by kitfoxer** [reply]
Is that in lieu of "sparse-color:" ?

**Reply by fmw42** [reply]
In the command that ends with txt:

**Reply by fmw42** [reply]
If you get the Unix tools for Windows - sed,  then you can filter the txt output just for the rgb values.


`magick rose:[5x5+0+0] -depth 8 txt: | sed -n 's/^.*(\(.*\)).*$/rgb(\1)/p'`


```
rgb(48,47,45)
rgb(50,48,46)
rgb(54,50,47)
rgb(56,51,46)
rgb(58,51,45)
rgb(47,46,44)
rgb(48,47,45)
rgb(53,48,45)
rgb(55,49,45)
rgb(56,49,44)
rgb(44,44,44)
rgb(44,45,43)
rgb(48,47,44)
rgb(49,46,43)
rgb(50,44,40)
rgb(44,45,47)
rgb(43,45,47)
rgb(46,45,45)
rgb(47,45,41)
rgb(47,42,39)
rgb(44,47,50)
rgb(43,47,50)
rgb(46,46,48)
rgb(45,46,45)
rgb(45,43,41)
```

**Reply by fmw42** [reply]
See https://sourceforge.net/projects/unxutils/ for unix tools for windows
